src/java/org/apache/cassandra/utils/concurrent/NonBlockingRateLimiter.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils.concurrent;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Ticker;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A rate limiter implementation that allows callers to reserve permits that may only be available 
+ * in the future, delegating to them decisions about how to schedule/delay work and whether or not
+ * to block execution to do so.
+ */
+@SuppressWarnings("UnstableApiUsage")
+@ThreadSafe
+public class NonBlockingRateLimiter
+{
+    private static final long DEFAULT_BURST_NANOS = TimeUnit.SECONDS.toNanos(1);
+    
+    /** a starting time for elapsed time calculations */
+    private final long startedNanos;
+    
+    /** 
+     * nanoseconds of "burst time"
+     * 
+     * ex. If this is configured at the default of 1 second, the limiter will allow a
+     *     number of permits equal to the configured number of permits/second to be issued 
```

Comment:
I'm about 75% sure it doesn't allow permits/second count, but instead allows for burstNanos / interval count to be allocated. For example, if nextAvailable is some time far in the past, you allocated at nowNanos - burstNanos, setting the next alloc at that sentinel + interval. Assuming a microburst of requests, your next will alloc at new sentinel + interval (nextAvailable == nowNanos - burstNanos), then add interval. Rinse and repeat in increments of interval (i.e. burstNanos as your original window / interval) until you cross nowNanos and get throttled.

Which may very well be fine. But _if_ that's how the logic is working, we should probably update the comment here to explain that.

Reply:
`interval = 1 / (permits/second)`, so if the burst is 1 second, using your expression, the number allocated is `1 / (1 / (permits/second))`, which I think is just `permits/second`. I think my mistake here was using that specific example and not describing the general case, which you are absolutely correct about. (If there is going to be an example, it should be after the general description as an illustration.)

Does that make sense?

Reply:
Ah, yes. That's a fairly clean way to describe the algorithm around allowing bursting; it's somewhat ironic that the long-winded example I gave above actually distills to `burstSeconds / (1 / (permits/second))`, or essentially just burstSeconds * permits/second. 👍 on updating the example in the comment for posterity. Thanks for the clarification and patience.